src/flash.rs

```diff
@@ -2,6 +2,242 @@
+    Sz32K  = 32,
+    Sz64K  = 64,
+    Sz128K = 128,
+    Sz256K = 256,
+    Sz384K = 384,
+    Sz512K = 512,
+    Sz768K = 768,
+    Sz1M   = 1024,
+}
+impl FlashSize {
+    fn size(self) -> u32 {
+        SZ_1K as u32 * self as u32
+    }
+}
+
+pub struct FlashWriter<'a> {
+    flash: &'a mut Parts,
+    sector_sz: SectorSize,
+    flash_sz: FlashSize,
+}
+impl<'a> FlashWriter<'a> {
+    fn unlock(&mut self) -> Result<()> {
+        // Wait for any ongoing operations
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        // NOTE(unsafe) write Keys to the key register
+        unsafe { self.flash.keyr.keyr().write(|w| w.key().bits(KEY1)); }
+        unsafe { self.flash.keyr.keyr().write(|w| w.key().bits(KEY2)); }
+
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        // Verify success
+        match self.flash.cr.cr().read().lock().bit_is_clear() {
+            true => Ok(()),
+            false => Err(Error::UnLockError),
+        }
+    }
+
+    fn lock(&mut self) -> Result<()> {
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        self.flash.cr.cr().modify(|_, w| w.lock().set_bit());
+
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        match self.flash.cr.cr().read().lock().bit_is_set() {
+            true => Ok(()),
+            false => Err(Error::LockError),
+        }
+    }
+
+    fn valid_address(&self, offset: u32) -> Result<()> {
+        if FLASH_START + offset > FLASH_END {
+            Err(Error::AddressLargedThanFlash)
+        }
+        else if offset % 2 != 0 {
+            Err(Error::AddressMisAligned)
+        }
+        else {
+            Ok(())
+        }
+    }
+
+    fn valid_length(&self, length: u32) -> Result<()> {
+        if length > self.sector_sz.size() as u32 {
+            Err(Error::LengthTooLong)
+        }
+        else if length % 2 != 0 {
+            Err(Error::LengthNotMultiple2)
+        }
+        else {
+            Ok(())
+        }
+    }
+
+    pub fn page_erase(&mut self, start_offset: u32) -> Result<()> {
+        self.valid_address(start_offset)?;
+
+        self.unlock()?;
+
+        // Set Page Erase
+        self.flash.cr.cr().modify(|_, w| w.per().set_bit() );
+
+        // Write address bits
+        // NOTE(unsafe) This sets the sector address in the Address Register
+        unsafe { self.flash.ar.ar().write(|w| w.far().bits(FLASH_START + start_offset) ); }
+
+        // Start Operation
+        self.flash.cr.cr().modify(|_, w| w.strt().set_bit() );
+
+        // Wait
+        while self.flash.sr.sr().read().bsy().bit_is_set() { }
+
+        // Check for errors
+        let sr = self.flash.sr.sr().read();
+
+        // Remove Page Erase Operation bit
+        self.flash.cr.cr().modify(|_, w| w.per().clear_bit() );
+
+        // Re-lock flash
+        self.lock()?;
+
+        for idx in 0..self.sector_sz.size() -1 {
+            let write_address = (FLASH_START + start_offset + idx as u32) as *mut u16;
```

Comment:
Don't you have to correct this to not go pass a section boundary ? i.e `start_offset` might be in the middle of a sector. Also I don't think it needs to be `mut`.